syntax = "proto2";
package naothmessages;
option java_package= "de.naoth.rc.core.messages";

import "CommonTypes.proto";
import "Framework-Representations.proto";

// Representations //

message CameraMatrix {
  required Pose3D pose = 1;
  optional CameraID cameraID = 2 [default=bottom];
  optional bool valid = 3;
  optional uint32 timestamp = 4;
}

message BallPercept {
  optional bool ballWasSeen = 1;
  optional DoubleVector2 centerInImage = 2;
  optional double radiusInImage = 3;
  optional Color ballColor = 4;
  optional DoubleVector2 bearingBasedOffsetOnField = 5;
  optional FrameInfo frameInfoWhenBallWasSeen = 6;
}

message MultiBallPercept {
  message BallPercept {
    optional DoubleVector2 centerInImage = 1;
    optional double radiusInImage = 2;
    optional DoubleVector2 positionOnField = 3;
    optional CameraID cameraId = 4;
  }

  optional FrameInfo frameInfoWhenBallWasSeen = 1;
  repeated BallPercept percepts = 2;
}

message BallModel {
  optional bool valid = 1;
  optional bool knows = 2;

  optional DoubleVector2 position = 3;
  optional DoubleVector2 speed = 4;

  optional DoubleVector2 positionPreview = 5;
  optional DoubleVector2 positionPreviewInLFoot = 6;
  optional DoubleVector2 positionPreviewInRFoot = 7;
  repeated DoubleVector2 futurePosition = 8 [deprecated=true];

}

message GoalPercept {
  message GoalPost {
    enum PostType
    {
      rightPost = 0;
      leftPost = 1;
      unknownPost = 2;
    };

    optional IntVector2 basePoint = 1;
    optional DoubleVector2 position = 2;
    optional Color color = 3;
    optional PostType type = 4;
    optional bool positionReliable = 5;
    optional double seenHeight = 6;
  }

  optional double angleToSeenGoal = 1;
  optional DoubleVector3 goalCentroid = 2;
  optional int32 numberOfSeenPosts = 3;
  repeated GoalPost post = 4;
}

message HeadMotionRequest {
  required uint32 id = 1;
  required uint32 cameraID = 2;
  optional DoubleVector2 targetJointPosition = 3;
  optional DoubleVector2 targetPointInImage = 4;
  optional DoubleVector3 targetPointInTheWorld = 5;
  optional DoubleVector3 searchCenter = 6;
  optional DoubleVector3 searchSize = 7;
  optional bool searchDirection = 8;
  optional DoubleVector2 targetPointOnTheGround = 9;
  optional uint32 coordinate = 10;
  optional double velocity = 11;
}

message StepControlRequest
{
  enum StepType {
	WALKSTEP = 0;
	KICKSTEP = 1;
	ZEROSTEP = 2;
  };

  required uint32 stepID = 1;
  required bool moveLeftFoot = 2;
  required Pose2D target = 3;
  required uint32 time = 4;
  required double speedDirection = 5;
  required double scale = 6;
  required StepType steptype = 7;
  
  enum RestrictionMode {
	HARD = 0;
	SOFT = 1;
  };
  
  required RestrictionMode restriction = 8;
  required bool isProtected = 9;
  required uint32 stepRequestID = 10;
}

message WalkRequest {
  required uint32 coordinate = 1;
  required double character = 2;
  required Pose2D target = 3;
  optional StepControlRequest stepControl = 4;
  required Pose2D offset = 5;
}

message KickRequest {
  required DoubleVector3 kickPoint = 1;
  required double kickDirection = 2;
  required int32 kickFoot = 3;
  required bool finishKick = 4;
}

message GraspRequest {
  required DoubleVector3 graspingPoint = 1;
  required uint32 graspingState = 2;
  required uint32 graspDistState = 3;
  required uint32 graspStiffState = 4;
}

message ArmMotionRequest {
  required uint32 id = 1;
  optional DoubleVector2 lShoulderPosition = 2;
  optional DoubleVector2 lShoulderStiffness = 3;
  optional DoubleVector2 lElbowPosition = 4;
  optional DoubleVector2 lElbowStiffness = 5;
  optional DoubleVector2 rShoulderPosition = 6;
  optional DoubleVector2 rShoulderStiffness = 7;
  optional DoubleVector2 rElbowPosition = 8;
  optional DoubleVector2 rElbowStiffness = 9;
}

message MotionRequest {
  required uint32 id = 1;
  required uint32 time = 2;
  required bool forced = 3;
  optional WalkRequest walkRequest = 4;
  optional bool starndardStand = 5;
  optional KickRequest kickRequest = 6;
  optional double standHeight = 7;
  optional bool calibrateFootTouchDetector = 8;
  optional uint32 cognitionFrameNumber = 9;
  optional GraspRequest graspRequest = 10;
  optional ArmMotionRequest armMotionRequest = 11;
  optional bool disable_relaxed_stand = 12;
}


message LinePercept {

  message Intersection {

    enum IntersectionType
    {
      unknown = 0;
      T = 1;
      L = 2;
      C = 3;
      none = 4;
      X = 5;
    }

    optional DoubleVector2 posInImage = 1;
    optional DoubleVector2 posOnField = 2;

    optional IntersectionType type = 3;
    optional uint32 segmentOneIndex = 6;
    optional uint32 segmentTwoIndex = 7;
    optional double segmentOneDistance = 8;
    optional double segmentTwoDistance = 9;
  }

  message FieldLineSegment {
    optional LineSegment lineInImage = 1;
    optional LineSegment lineOnField = 2;
    optional int32 beginExtendCount = 3;
    optional int32 endExtendCount = 4;
    optional double thickness = 5;
    optional double slope = 6;
    optional double angle = 7;

    optional bool valid = 8;
  }

  repeated FieldLineSegment lines = 1;
  repeated Intersection intersections = 2;
  optional DoubleVector2 middleCircleCenter = 3;
}

message ShortLinePercept {
  repeated LineSegment fieldLineSegments = 1;
}

message Polygon {
  repeated IntVector2 points = 1;
}

message FieldPercept {
  required Polygon fieldPoly = 1;
}

message RansacLinePercept {
  repeated LineSegment fieldLineSegments = 1;
}

message RansacCirclePercept2018 {
  optional bool wasSeen = 1;
  optional DoubleVector2 middleCircleCenter = 2;

}

message ScanLineEndPoint {
  required IntVector2 posInImage = 1;
  required DoubleVector2 posOnField = 2;
  required Color color = 3;
  required uint32 ScanLineID = 4;
}

message DoubleEdgel {
  required DoubleVector2 point = 1;
  required DoubleVector2 direction = 2;
}

message Edgel {
  required IntVector2 point = 1;
  required DoubleVector2 direction = 2;
}

message EdgelPair {
  required int32 begin = 1;
  required int32 end = 2;
  required int32 id = 3;
  optional bool adaptive = 4;
}

message ScanLineEdgelPercept {
  repeated Edgel edgels = 1;
  repeated EdgelPair pairs = 3;
  repeated ScanLineEndPoint endPoints = 2;
}

message LineGraphPercept {
  repeated Edgel edgelsOnField = 1;
  
  repeated DoubleEdgel edgelsInImageTop = 2;
  repeated DoubleEdgel edgelsInImage = 3;
}

message StepControlStatus
{
  required uint32 stepID = 1;
  required uint32 moveableFoot = 2;
  required uint32 stepRequestID = 3;
}

message MotionStatus
{
  required uint32 time = 1;
  required uint32 lastMotion = 2;
  required uint32 currentMotion = 3;
  required uint32 headMotion = 4;
  required uint32 currentMotionState = 5;
  required Pose2D plannedMotionLeftFoot = 6;
  required Pose2D plannedMotionRightFoot = 7;
  required Pose2D plannedMotionHip = 8;
  optional StepControlStatus stepControlStatus = 9;
  
  optional bool   target_reached = 10;
  optional bool   head_target_reached = 11;
  optional bool   head_got_stuck = 12;
  
  optional bool   walk_emergency_stop = 13;
  
  optional bool   head_at_rest = 14;
}

message OdometryData
{
  required Pose2D pose = 1;
}

message BodyState
{
  optional bool isLiftedUp = 1;
}

message BodyStatus
{
  repeated double currentSum = 1;
  //repeated double currentSumFiltered = 2;
  required uint32 timestamp = 3;
}

message CalibrationData
{
  optional DoubleVector2 inertialSensorOffset = 1;
  //optional DoubleVector3 inertialSensorOffset = 6;
  //optional DoubleVector2 gyroSensorOffset = 2;
  optional DoubleVector3 gyroSensorOffset = 5;
  optional DoubleVector3 accSensorOffset = 3;
  required bool calibrated = 4;
}

message InertialModel
{
  optional DoubleVector2 orientation = 1;
}


message CameraMatrixCalibration
{
  repeated DoubleVector2 correctionOffset = 1 [deprecated=true];
  optional DoubleVector2 correctionOffsetBody = 2;
  optional DoubleVector3 correctionOffsetHead = 3;
  repeated DoubleVector3 correctionOffsetCam  = 4;
}

message BallCandidates
{
  message Patch
  {
    enum Type {
      Y = 0;
      YUV = 1;
      YUVC = 2;
    }
    required IntVector2 min = 1;
    required IntVector2 max = 2;
    optional Type type = 3 [default=Y];
    required bytes data = 5;
  }

  repeated Patch patches = 1;
}

message RemoteControlCommand {
  enum ControlMode
  {
    DIRECT_CONTROL = 0;
    LOCK_CONTROL = 1;
  }

  enum ActionType {
    NONE = 0;
    STAND = 1;
    WALK = 2;
    KICK_RIGHT = 3;
    KICK_LEFT = 4;
    KICK_FORWARD_LEFT = 5;
    KICK_FORWARD_RIGHT = 6;
  }

  enum SecondActionType
  {
    SECOND_NONE = 0;
    BLINK = 1;
    SAY = 2;
  }

  optional ControlMode controlMode = 1;
  optional ActionType action = 2;
  optional SecondActionType second_action = 3;

  optional Pose2D target = 4;
}

message GroundContactModel {
  enum Foot {
    LEFT = 0;
    RIGHT = 1;
    NONE = 2;
  }

  required bool leftGroundContact = 1;
  required bool rightGroundContact = 2;
  required Foot supportFoot = 3;
}


message CollisionPercept {
  required uint32 timeCollisionArmLeft = 1;
  required uint32 timeCollisionArmRight = 2;
  required bool isCollisionLeftBumper = 3;
  required bool isCollisionRightBumper = 4;
  required double lastComputedCollisionLeft = 5;
  required double lastComputedCollisionRight = 6;
}

message IMUData
{
 required DoubleVector3 location = 1;
 required DoubleVector3 velocity = 2;
 required DoubleVector3 acceleration = 3;
 optional DoubleVector3 acceleration_sensor = 4;

 required DoubleVector3 rotation = 5;
 required DoubleVector3 rotational_velocity = 6;
 optional DoubleVector3 rotational_velocity_sensor = 7;
 required DoubleVector2 orientation = 8;
 required DoubleVector3 orientation_rotvec = 9;

 optional bool has_been_reset = 10;
}

message CentreOfPressure {
 required DoubleVector3 in_and_only_left_foot_cop = 1;
 required double in_and_only_left_foot_magnitude  = 2;
 required bool   in_and_only_left_foot_valid      = 3;

 required DoubleVector3 in_and_only_right_foot_cop = 4;
 required double in_and_only_right_foot_magnitude  = 5;
 required bool   in_and_only_right_foot_valid      = 6;

 required DoubleVector3 in_kinematic_chain_origin_cop = 7;
 required double in_kinematic_chain_origin_magnitude  = 8;
 required bool   in_kinematic_chain_origin_valid      = 9;
}

message CalibrationDataCMC {
 message CalibrationDataSampleV3 {
  optional Pose3D chestPose               = 1;
  repeated DoubleVector2 edgelsInImage    = 2;
  repeated DoubleVector2 edgelsInImageTop = 3;
  optional DoubleVector2 orientation      = 4;
  optional double headYaw                 = 5;
  optional double headPitch               = 6;
 }

 repeated CalibrationDataSampleV3 calibrationData = 1;
 optional uint32 numberOfResudials                = 2;
}

message WhistlePercept {
  message Whistle {
    required string name = 1;
    required int64 positionInCapture = 2;
    required double responseValue = 3;
  }
  
  optional uint32 frameWhenWhistleDetected  = 1;
  optional string captureFile = 2;
  repeated Whistle recognizedWhistles = 3;
  optional bool whistleDetected = 4;
}

message DebugModify {
  message ModifyValue {
    required string name = 1;
    optional bool modify = 2;
    optional double value = 3;
  }
  
  repeated ModifyValue valueMap = 1;
}


message StepBuffer {
  optional Pose3D support_foot = 1;
}

message bodyAwareness {
  optional bool isLifted = 1;
  optional DoubleVector3  calculatedCoM= 2;
  optional DoubleVector3 currentCoMError = 3;
}
